timemachine/fe/stored_arrays.py

```diff
@@ -0,0 +1,140 @@
+import io
+import shutil
+import tempfile
+import weakref
+from itertools import count
+from pathlib import Path
+from typing import Iterator, List, NoReturn, Sequence, overload
+
+import numpy as np
+from numpy.typing import ArrayLike, NDArray
+
+from timemachine.parallel.client import AbstractFileClient
+
+
+class StoredArrays(Sequence[NDArray]):
+    """Sequence of numpy arrays using O(1) memory, backed by disk storage.
+
+    Data is stored in a temporary directory that is cleaned when the `StoredArrays` object is finalized.
+
+    Examples
+    --------
+    >>> sa = StoredArrays()
+    >>> sa.extend([np.array([1, 2, 3]), np.array([4, 5, 6])])
+    >>> len(sa)
+    2
+    >>> len(list(sa._chunks()))
+    1
+    >>> sa.extend([np.array([7, 8, 9])])
+    >>> len(sa)
+    3
+    >>> len(list(sa._chunks()))
+    2
+    >>> list(sa)
+    [array([1, 2, 3]), array([4, 5, 6]), array([7, 8, 9])]
+    """
+
+    def __init__(self):
+        self._chunk_sizes: List[int] = []
+        self._dir = Path(tempfile.mkdtemp())
+        self._finalizer = weakref.finalize(self, shutil.rmtree, self._dir)
+
+    def __iter__(self) -> Iterator[NDArray]:
+        for chunk in self._chunks():
+            for array in chunk:
+                yield array
+
+    def __len__(self) -> int:
+        return sum(self._chunk_sizes)
+
+    @overload
+    def __getitem__(self, key: int) -> NDArray:
+        ...
```

Comment:
Should these raise exceptions rather then return None?

Reply:
what exactly are these doing? Seems like `__getitem__` below handles these cases.

Reply:
The `@overload` decorated methods are stubs (purely for type checking; see https://docs.python.org/3/library/typing.html#typing.overload)